### PR TITLE
spike: capacitor 4 upgrade

### LINKED ISSUE #1553: [FEATURE] Android/Capacitor Core Updates
**What?**
Capacitor and android core deps require updating

**Why?**
Periodically android requires updates to platforms targetted or security patches to packages used. Our android code is coupled tightly with capacitorjs, so it makes sense to update both at the same time (capacitor v3->v4 and various android file changes)

E.g. Issues raised when deploying app (first 3 should be addressed with updates)
image

**How?**
Follow instructions at: https://capacitorjs.com/docs/updating/4-0 to upgrade capacitor v3 -> v4 and core android updates
Check working as expected on android

Can also compare against a clean capacitor app to see if there's any legacy android code floating around

**Other details**
Can use the appetize emulator to test. Also important to ensure internal testing done on device before wider deployment.
It would probably also be a good time to update gradle settings as prompted from android studio (v4.x -> v7.x)

The most important things to test for will be app launch and native integrations (e.g. push notifications, splash screen, firebase analytics/performance/crashlytics)

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [ ] - Latest `master` branch merged
- [ ] - PR title descriptive (can be used in release notes)

## Description
- Update capcitor-cli to v4 and run migration script
- Review https://capacitorjs.com/docs/updating/4-0 (no additional changes made, appears all handled via cli migration)
- Check to ensure Java 11 available in CI (it is)
- Fix init issue caused by migration script
- Fix notifications id issue identified during upgrade

## Dev Notes
Issue noted of test notifications, says that they are scheduled but won't allow sending immediately 
image

Assume it's caused from `https://github.com/ionic-team/capacitor-plugins/pull/1303/files`, turns out that notification ids have a value limit which we aren't respecting. Also fixed in pr

## Git Issues

Closes #1553

## Screenshots/Videos
Migration output - some wrapper files failed (assume non-issue as gradle can repopulate)
```
√ Migrating AndroidManifest.xml by adding android:exported attribute to Activity. in 1.31ms
[?] Some projects still require JCenter to function. If your project does, please answer yes below.
√ Keep JCenter if present? (y/N) ... y
[info] Set com.android.tools.build:gradle = 7.2.1.
[info] Set com.google.gms:google-services = 4.3.13.
[info] Added mavenCentral().
[info] Added mavenCentral().
√ Migrating build.gradle file. in 14.50ms
√ Migrating app/build.gradle file. in 1.07ms
√ Migrating gradle-wrapper.properties by updating gradle version from 7.0 to 7.4.2. in 789.00μs
√ Migrating variables.gradle file. in 8.43ms
√ Migrating MainActivity in 6.45ms
√ Migrate to Android 12 Splashscreen and apply DayNight theme. in 969.20μs
√ Running cap sync. in 8.03s× Upgrading gradle wrapper files - failed![error] gradle wrapper files were not updated
[info] IMPORTANT: Review https://capacitorjs.com/docs/updating/4-0#plugins for breaking changes in these plugins       that you use: @capacitor/push-notifications, @capacitor/local-notifications.
[info] Warning: The Android Gradle plugin was updated and it requires Java 11 to run. You may need to select
       this in Android Studio.
√ Writing breaking changes. in 1.25ms
[success] Migration to Capacitor ^4.0.0 is complete. Run and test your app!
```

**Tested notification working**
image

### CODE REVIEW

**Review by jfmcquade — CHANGES_REQUESTED** [review]
## Capacitor Migration
I've cross-referenced the changes with those in the migration guide [here](https://capacitorjs.com/docs/updating/4-0).
All the steps from the guide appear to have been completed correctly, apart from one, which i've flagged in a comment (and therefore requested changes).

## Testing (using appetize build)
### Notifications
Notifications seem to work as they should when I use the "send now" feature.

I did experience an issue with notifications when I first fired up the appetize preview: I immediately received 4 notifications. These appear to be the 4 notifications that are scheduled erroneously (NB for 1pm) as noted in issue #1640  
From appetize:
Screenshot 2022-12-05 at 13 00 54
From local master build:
Screenshot 2022-12-05 at 15 22 42

They should be cleared from the scheduled notifications whenever another notification is scheduled.

When I reload the appetize preview page, I can't see any scheduled notifications that match the ones I received. As it happens, I was testing at exactly 1pm, so I think that the notifications were scheduled and sent during the short space of time that the app was initialising, where usually they would be unscheduled before they could be sent. I haven't been able to recreate this bug. Indeed, the fact that those unwanted scheduled notifications do not appear in the appetize build suggests that the bug is not extant in the build.

Testing with the workshop skin activated appears to confirm that the campaign notifications are being scheduled correctly:
appetize build:
Screenshot 2022-12-05 at 15 33 44
local master build:
Screenshot 2022-12-05 at 15 26 51

### Android assets
In the appetize preview, the splash screen and app icon both appear as expected.

### Google Sign-In
The TZ deployment is the one that does enforce Google sign-in (via the `APP_AUTHENTICATION_DEFAULTS.enforceLogin` config option). So it is important that the sign-in functionality is working.
The appetize build of this PR is using the `plh_za` deployment (as shown by the default app language at the bottom of this screenshot of the build)
Screenshot 2022-12-05 at 16 04 02
I'm not sure how to configure it to build using the `plh_tz` deployment, which would be the only way to fully test the google sign-in functionality.

As it is, I can navigate to the `sign_in` template manually, and can confirm that it correctly prompts me to sign in with a google account (or select an existing google account if there is already one signed in to the emulated device). There is no confirmation from this template that the sign in is successful, but none is expected. See video below
https://user-images.githubusercontent.com/64838927/205682808-3bee7a79-5366-4dc7-ab63-b78be0827685.mov

File: android/app/src/main/java/international/idems/plh_teens/MainActivity.java [hunk 1]
```diff
@@ -12,6 +12,7 @@ public class MainActivity extends BridgeActivity {
   @Override
   public void onCreate(Bundle savedInstanceState) {
     super.onCreate(savedInstanceState);
+    // https://github.com/baumblatt/capacitor-firebase-auth/issues/170
     registerPlugin(CapacitorFirebaseAuth.class);
```
[the remark this answers is not in the document]
Comment (chrismclarke) [reply android/app/src/main/java/international/idems/plh_teens/MainActivity.java:16]: thanks, that was the bit of code where the cli actually added super.oncreate both before and after the PR causing all sorts of build errors which were a pain to track down. So I think in the end I reverted back to the original build settings, but given they make an explicit point of it makes sense to swap

**Comment by chrismclarke** [comment]
Thanks for checking everything @jfmcquade - updated the java imports as suggested so hopefully good to go.

Very strange about the notifications, although as there's existing issues for it (and a few of the draft prs also affect notifications) I'm guessing best to resolve separately if it comes up again.

One additional minor thing I noticed in the docs was that from Android 12 extra permissions required to schedule exact notification. I'm assuming we don't need this (designed more for alarms) and likely the notification will come through around that time
https://capacitorjs.com/docs/apis/local-notifications#android

**Comment by chrismclarke** [comment]
Oh, and as for google sign-in I double checked crashlytics and no errors reported there, also usually if something were to fail it would be before the google sign-in prompt so I'd say all good

**Review by jfmcquade — APPROVED** [review]
All looks good

**Pull request merged**